Implement `CLIENT KILL MAXAGE <maxage>`

slavak opened the pull request:
Adds an ability to kill clients older than a specified age.

Also, fixed the age calculation in `catClientInfoString` to use `commandTimeSnapshot`
instead of the old `server.unixtime`, and added missing documentation for
`CLIENT KILL ID` to output of `CLIENT help`.

oranagra commented on src/networking.c:
```diff
@@ -3179,6 +3196,7 @@ NULL
             if (id != 0 && client->id != id) continue;
             if (user && client->user != user) continue;
             if (c == client && skipme) continue;
+            if (max_age != 0 && (long long)(server.unixtime - client->ctime) < max_age) continue;
```
i guess we better use `commandTimeSnapshot`

oranagra commented on tests/unit/introspection.tcl:
```diff
@@ -61,6 +65,26 @@ start_server {tags {"introspection"}} {
+            set rd2 [redis_deferring_client]
+            set connected_clients [s connected_clients]
+            assert {$connected_clients >= 3}
+            # Should kill everything except r and rd2
+            set res [r client kill age 1]
```
why won't it kill `r`?
p.s. we do have an implicit `r reconnect` feature, but i hate to rely on it, so i'd rather not disconnect `r`.
we can use `skipme`, but maybe we'll combine this with the `user` option? i.e. make both rd1 and rd2 use a specific user?

slavak replied:
Because `SKIPME` defaults to yes.

We could probably add a user-based filter if you're worried. I was trying to keep it simple.

oranagra replied:
Ohh. I'm ok keeping it as is, but maybe an explicit SKIPME or a USER can make it more readable

oranagra replied:
Actually, now that I look at it. If we kill the only two clients that are both old and use that user so we can't be sure if the age works. We need another one that uses the same user but doesn't match the age criteria.

slavak replied:
I think you've gotten a bit turned around. :slightly_smiling_face:  The test kills `rd1` _but not_ `rd2`.

I did change it to use a dummy user, and also just added a small fix to the test to account for that.

oranagra replied:
sorry, was reviewing it from android, and the comment was misleading..
looks good now.
p.s. please avoid force-pushes.

oranagra commented on tests/unit/introspection.tcl:
```diff
@@ -61,6 +65,26 @@ start_server {tags {"introspection"}} {
+            set res [r client kill age 1]
+            assert {$res == $connected_clients - 2}
+
+            # rd2 should still be connected
+            $rd2 ping
+            assert_equal "PONG" [$rd2 read]
```
there's an implicit reconnect, so i'm not sure that's a valid test.
maybe we should use `client setname`, and check that it's missing from `client list`?

slavak replied:
I stole the logic from the `CLIENT KILL SKIPME` test (the one immediately following this one), so I was kinda hoping it was correct.

(It does work, FWIW; the test fails without the new code, and passes with it.)

oranagra commented on tests/unit/introspection.tcl:
```diff
@@ -61,6 +65,26 @@ start_server {tags {"introspection"}} {
+            $rd1 close
+            $rd2 close
+        }
+} ;# tag
```
you can use this pattern:
```
test {name} {
    code
} {} {tag}
```
i.e. the last two arguments are expected return value and tag.

slavak commented:
@oranagra Pushed a new commit with CR fixes:

* Use `commandTimeSnapshot` instead of `server.unixtime`. (Also fixed in the existing code for returning client age, such as for `CLIENT LIST`.)
* Used the suggested tag syntax for the unit test.
* Tweaked the unit test to filter on user as an extra precaution.

I updated the top PR comment so it's correct when squash-merging.

oranagra commented:
@redis/core-team please approve

madolson commented on src/networking.c:
```diff
@@ -3042,6 +3042,10 @@ void clientCommand(client *c) {
 "      Skip killing current connection (default: yes).",
+"    * ID <client-id>",
+"      Kill connections by client id.",
+"    * AGE <max-age>",
```
```suggestion
"    * MAXAGE <max-age>",
```
Any reason to not be specific here?

oranagra commented:
approved in core-team meeting with the above comment (AGE -> MAXAGE)

oranagra commented:
@slavak finally merged.
wanna do the redis-doc PR?
